src/library/crate/cratetablemodel.h

```diff
@@ -1,37 +1,30 @@
-#ifndef MIXXX_CRATETABLEMODEL_H
-#define MIXXX_CRATETABLEMODEL_H
-
+#pragma once
 
 #include "library/basesqltablemodel.h"
-
 #include "library/crate/crateid.h"
 
-
 class CrateTableModel : public BaseSqlTableModel {
     Q_OBJECT
 
   public:
     CrateTableModel(QObject* parent, TrackCollectionManager* pTrackCollectionManager);
     ~CrateTableModel() final;
 
-    void selectCrate(
-        CrateId crateId = CrateId());
-    CrateId selectedCrate() const {
+    void selectCrate(CrateId crateId = CrateId());
+    [[nodiscard]] CrateId selectedCrate() const {
```

Comment:
We don't use this anywhere else. Now that we switched to C++17 it might be a good idea to start using it for simple getter functions like this. @uklotzde @daschuer Opinions?

Reply:
It improves safety.

On the other hand it introduces a lot of noise to the code and makes it harder to read. Functions return results for a reason and results should never be discarded. If the result is not mandatory it should be `void`, indicating that the "purpose" of this function is to cause a side-effect.

Personally I would instead like to enable a warning for **all** discarded return values by default. The caller should be responsible to explicitly **opt-out** for every discarded return value. Considering the amount of legacy code this might not be feasible.

Reply:
just followed the instructions of clang-tidy here - maybe we can configure that as well to fit your style?

Reply:
This is just a simple, idempotent getter method. I don't see any value in introducing this attribute here.

An agreed set of rules for clang-tidy would be preferable. Otherwise we will argue repeatedly where a certain style is appropriate. The same situation already applies to many other informal C++ conventions that are not checked by the compiler.

Reply:
Reading here: https://en.cppreference.com/w/cpp/language/attributes/nodiscard
that new() has this attribute. It is clear that a bad thing = "memory leak" happens when the return value is discarded. This is obvious during writing the called function. 
I think it is reasonable to follow this pattern.  

The crucial point here is if, this applies to simple getter functions. If a user discards the value,
I guess the compiler will discard the whole call. So nothing bad happens form the called function perspective. However it is  most likely an issue from the caller perspective. 

I think I would not use it here in favor of less boilerplate.

Reply:
> The crucial point here is if, this applies to simple getter functions. If a user discards the value,
I guess the compiler will discard the whole call. So nothing bad happens form the called function perspective.

It will likely be discarded (hehe) by the compiler, but it clearly indicates bad code. I'd be in favor of using it.

Reply:
But using it only here and not everywhere else and especially in places where it actually would be more appropriate as Daniel explained is questionable.

Reply:
This is what `-Wunused-result` is for instead of polluting the code. The attribute is only appropriate where ignoring the return value is an error from the perspective of the callee, i.e. by causing memory leaks. Otherwise the function itself should not need to care by whom it is called and how the result is used or not.